lib/mqtt.c

```diff
@@ -0,0 +1,481 @@
+
+#ifndef CURL_DISABLE_MQTT
+
+#include "urldata.h"
+#include <curl/curl.h>
+#include "transfer.h"
+#include "sendf.h"
+#include "progress.h"
+#include "mqtt.h"
+#include "select.h"
+#include "strdup.h"
+#include "url.h"
+#include "escape.h"
+#include "warnless.h"
+#include "curl_printf.h"
+#include "curl_memory.h"
+#include "multiif.h"
+
+/* The last #include file should be: */
+#include "memdebug.h"
+
+#define MQTT_MSG_CONNECT   0x10
+#define MQTT_MSG_PUBLISH   0x30
+#define MQTT_MSG_SUBSCRIBE 0x82
+
+#define MQTT_CONNACK_LEN 4
+#define MQTT_SUBACK_LEN 5
+
+/*
+ * Forward declarations.
+ */
+
+static CURLcode mqtt_do(struct connectdata *conn, bool *done);
+static CURLcode mqtt_doing(struct connectdata *conn, bool *done);
+static int mqtt_getsock(struct connectdata *conn,
+                        curl_socket_t *sock, int numsocks);
+
+/*
+ * MQTT protocol handler.
+ */
+
+const struct Curl_handler Curl_handler_mqtt = {
+  "MQTT",                             /* scheme */
+  ZERO_NULL,                          /* setup_connection */
+  mqtt_do,                            /* do_it */
+  ZERO_NULL,                          /* done */
+  ZERO_NULL,                          /* do_more */
+  ZERO_NULL,                          /* connect_it */
+  ZERO_NULL,                          /* connecting */
+  mqtt_doing,                         /* doing */
+  ZERO_NULL,                          /* proto_getsock */
+  mqtt_getsock,                       /* doing_getsock */
+  ZERO_NULL,                          /* domore_getsock */
+  ZERO_NULL,                          /* perform_getsock */
+  ZERO_NULL,                          /* disconnect */
+  ZERO_NULL,                          /* readwrite */
+  ZERO_NULL,                          /* connection_check */
+  PORT_MQTT,                          /* defport */
+  CURLPROTO_MQTT,                     /* protocol */
+  PROTOPT_NONE                        /* flags */
+};
+
+static CURLcode mqtt_busy_write(struct connectdata *conn,
+                                char *buf, size_t len)
+{
+  CURLcode result = CURLE_OK;
+  curl_socket_t sockfd = conn->sock[FIRSTSOCKET];
+
+  while(len > 0) {
+    ssize_t n;
+    result = Curl_write(conn, sockfd, buf, len, &n);
+    if(result && result != CURLE_AGAIN)
+      break;
+    buf += n;
+    len -= n;
+  }
+  return result;
+}
+
+/* Generic function called by the multi interface to figure out what socket(s)
+   to wait for and for what actions during the DOING and PROTOCONNECT states*/
+static int mqtt_getsock(struct connectdata *conn,
+                        curl_socket_t *sock, /* points to numsocks
+                                                number of sockets */
+                        int numsocks)
+{
+  (void)conn;
+  (void)numsocks;
+  sock[0] = conn->sock[FIRSTSOCKET];
+  return GETSOCK_READSOCK(FIRSTSOCKET);
+}
+
+static CURLcode mqtt_connect(struct connectdata *conn)
+{
+  CURLcode result = CURLE_OK;
+  const char clientid[] = "curl1234abcd";
```

Comment:
Two users will not be able to use curl concurrently if static client id is used. According to spec, client id must be unique so existing client will be disconnected.

Reply:
This variable is actually only used to find out the length of the client id string (rather than using a magic number). The real client id sent is random and written into the packet a few lines below:
`msnprintf(packet + client_id_offset, sizeof(packet), "curl%08x", rand());`

But this code isn't terribly elegant, and your misunderstanding shows me I should rewrite it to be more clear. Thanks for reviewing!